### PR TITLE
Add a non- PyTorch Lightning ``Trainer``

### PR DESCRIPTION
**Pull request by nathancooperjones**
## Describe the big picture - why are we doing this?
Through some sick, twisted fate, it turns out that PyTorch Lightning is actually slower than a vanilla PyTorch training loop. In order to better speed up our training jobs, I've added another trainer object that has a similar API to PyTorch Lightning's `Trainer`, but actually just implements the standard PyTorch training loop! 

Since Michael is reviewing this, nothing here should look too out-of-the-ordinary, except maybe some of the logging we do so we can continue to use PyTorch Lightning logging. Any input here on how to speed this training loop up any more / things that are here that really shouldn't is appreciated! 

## Any additional details to clarify code in the PR?
The bulk of this PR is in `collie_recs/model/base.py`. Just about every other code change here is slight renamings of things. 

## How is this tested?
I've added some monster tests in `tests/test_model.py`. Every line of the `CollieTrainerNoLightning` is tested in at least one of these new tests, so our code coverage is still high! 

## Pull Request Checklist
 - [X] Pull request includes a description of why we are doing this
 - [X] ``CHANGELOG`` has been updated
 - [X] Version in ``_version.py`` has been updated
 - [X] All tests in the ``tests`` folder pass with a local build
 - [ ] ``README`` has been updated (if applicable)
 - [X] Documentation in ``docs`` has been updated (if applicable)
 - [X] ``requirements.txt`` and ``requirements-dev.txt`` have been recompiled (if applicable)

### CODE REVIEW

File: collie_recs/model/base.py [hunk 1]
```diff
@@ -91,6 +144,350 @@ def __init__(self,
...
+    addition of the additional ``Trainer`` arguments and ``callbacks``, we designed this class as a
+    way to train a model in production, where we might be more focused on faster training times and
+    less on hyperparameter tuning and R&D, where one might instead opt to use PyTorch Lightning's
+    ``Trainer`` class.
+
+    Note that the arguments the ``CollieTrainerNoLightning`` trainer accepts will be slightly
+    different than the ones that the ``CollieTrainer`` accept, and defaults are also not guaranteed
+    to be equal as the two libraries evolve. Notable changes are:
+
+    * If ``gpus > 1``, only a single GPU will be used. Multi-GPU training is not supported in
```
Comment (sugi-chan) [inline collie_recs/model/base.py:169]: do the default behavior here would be pytorch would allocate the memory for the single GPU but the others would remain unused? idk if its worth noting the way in which Pytorch doesnt make use of the gpus more sepcifically lol

File: collie_recs/model/base.py [hunk 2]
```diff
@@ -91,6 +144,350 @@ def __init__(self,
         super().__init__(**kwargs)
 
 
+class CollieTrainerNoLightning():
```
[the remark this answers is not in the document]
Comment (sugi-chan) [reply collie_recs/model/base.py:147]:
```
CollieTrainerdeLigntninged
CollieTrainerSansLigntning
```
how to disperse lightning/electricity
```
GroundedCollieTrainer
CollieGroundedTrainer
CollieTrainerConductor
CollieConductorTrainer

```

Things like lightning?
```
CollieTrainerBolt
CollieBoltTrainer
CollieTrainerThunderLight
CollieThunderLightTrainer

```

words to think of? idk
```
firebolt or fireball #channel that dnd energy
earth? opposite 
```

File: collie_recs/model/base.py [hunk 3]
```diff
@@ -91,6 +144,350 @@ def __init__(self,
...
+        self._finalize_training()
+
+    def _pre_training_setup(self, model):
+        """Set up DataLoaders, optimizers, learning rate schedulers, etc. before training starts."""
+        self.train_dataloader = model.train_dataloader()
+        self.val_dataloader = model.val_dataloader()
+
+        self.lr_scheduler = None
+        configure_optimizers_return_value = model.configure_optimizers()
+        if isinstance(configure_optimizers_return_value, tuple):
```
Comment (sugi-chan) [inline collie_recs/model/base.py:383]: this is a cool way to be doing the setup for the optimizers and Lr schedulers

File: collie_recs/model/base.py [hunk 4]
```diff
@@ -91,6 +144,350 @@ def __init__(self,
...
+            # we have a list of optimizers and a list of lr_schedulers dictionaries
+            optimizers, lr_schedulers = configure_optimizers_return_value
+            self.optimizer = MultiOptimizer(optimizers)
+            self.lr_scheduler = MultiLRScheduler(lr_schedulers)
+        elif isinstance(configure_optimizers_return_value, list):
+            # we have a list of optimizers
+            self.optimizer = MultiOptimizer(configure_optimizers_return_value)
```
Comment (sugi-chan) [inline collie_recs/model/base.py:390]: so in this case how does collie build the lr scheduler bit. is it just a single one that gets defined somewhere outside this function? or is the idea that here there is no scheduler and it just goes since its set to None on 381?

File: collie_recs/cross_validation.py [hunk 5]
```diff
@@ -12,6 +12,29 @@
 from collie_recs.utils import get_random_seed
 
 
+def _subset_interactions(interactions: Interactions, idxs: Iterable[int]) -> Interactions:
```
Comment (sugi-chan) [inline collie_recs/cross_validation.py:15]: according to our favorite J.howard we dont need cross validation 🙄 

image

File: collie_recs/model/base.py [hunk 6]
```diff
@@ -42,6 +45,57 @@ def reset_parameters(self) -> None:
         self.weight.data.zero_()
 
 
+class MultiOptimizer(object):
```
Comment (sugi-chan) [inline collie_recs/model/base.py:48]: cool helpers with the optimizers and lr schedulers

File: collie_recs/model/base.py [hunk 7]
```diff
@@ -91,6 +144,350 @@ def __init__(self,
...
+            batch = self._move_batch_to_device(batch)
+            loss = model._calculate_loss(batch)
+            loss.backward()
+
+            self.optimizer.step()
+            self.train_steps += 1
+
+            detached_loss = loss.detach()
+
+            if self.terminate_on_nan and not torch.isfinite(loss).all():
```
Comment (sugi-chan) [inline collie_recs/model/base.py:434]: how often do you think you hit this? for some of your otherwork you had interesting solutions for this. so could be worth looking into if its an issue

File: collie_recs/model/base.py [hunk 8]
```diff
@@ -91,6 +144,350 @@ def __init__(self,
...
+
+        for batch_idx, batch in enumerate(self.val_dataloader):
+            batch = self._move_batch_to_device(batch)
+            loss = model._calculate_loss(batch)
+
+            self.val_steps += 1
+
+            total_loss += loss.detach()
+
+            if self.logger is not None:
```
Comment (sugi-chan) [inline collie_recs/model/base.py:464]: in my brain there is a way to combine this and the logging in the other one into a function or some shared code, but it seems largely unnecessary.

File: collie_recs/model/base.py [hunk 9]
```diff
@@ -91,6 +144,350 @@ def __init__(self,
...
+                if self.train_steps % self.flush_logs_every_n_steps == 0:
+                    self.logger.save()
+
+        return (total_loss / len(self.train_dataloader)).item()
+
+    def _val_loop_single_epoch(self, model):
+        """Validation loop for a single epoch, where gradients are NOT optimized for."""
+        total_loss = 0
+
+        for batch_idx, batch in enumerate(self.val_dataloader):
```
Comment (sugi-chan) [inline collie_recs/model/base.py:456]: Thinking out loud. is there any value in doing a loop where u have one thing vs the train and val and just dont do the loss backprop of its val?... probably not, id say leave as is

**Review by sugi-chan — COMMENTED** [review]
leaving general stuff for today... im still being largely unproductive, but the trainer stuff looks pretty good/cool so far. nicely done!

I still have to look through the tests/fixtures/that biz

File: collie_recs/model/base.py (hunk 4 above)
Comment (nathancooperjones) [reply collie_recs/model/base.py:390]: In this case, if we just have a list of optimizers, then we have no learning rate schedulers, so they'll stay at `None`.

File: collie_recs/cross_validation.py (hunk 5 above)
Comment (nathancooperjones) [reply collie_recs/cross_validation.py:15]: *deletes entire file*

File: collie_recs/model/base.py (hunk 7 above)
Comment (nathancooperjones) [reply collie_recs/model/base.py:434]: In Collie, I actually think I never hit this (unless the learning rate is set waaaayyyyy too high). I figured I would just implement it here since I like the option in PyTorch Lightning and it's not too much work.

File: collie_recs/model/base.py (hunk 9 above)
Comment (nathancooperjones) [reply collie_recs/model/base.py:456]: It's not super clear to me one way or another. 

I think I'm slightly more leaning towards keeping it as is, since there is not a ton of shared code between the two and having it all be a single loop would end up introducing a lot of extra complexities with a ton of `if/else` logic thrown in. I don't know, does that make sense?

Comment (sugi-chan) [reply collie_recs/model/base.py:458]: yeah I agree. its more of a thing where whenever i see these im like "what if..." but I think this setup makes sense

File: collie_recs/model/base.py (hunk 7 above)
Comment (sugi-chan) [reply collie_recs/model/base.py:438]: yeah that checks out

File: collie_recs/model/base.py (hunk 4 above)
Comment (sugi-chan) [reply collie_recs/model/base.py:396]: seems fair. how often do you train these without an LR scheduler?

Comment (nathancooperjones) [reply collie_recs/model/base.py:396]: Pretty much never, but maybe someone out there really, really hates LR schedulers and will appreciate this option?

File: collie_recs/model/base.py [hunk 10]
```diff
@@ -489,8 +897,8 @@ def training_epoch_end(
 
         self.log(name='train_loss_epoch', value=avg_loss)
 
-    def val_dataloader(self) -> Union[InteractionsDataLoader,
-                                      ApproximateNegativeSamplingInteractionsDataLoader]:
+    def val_dataloader(self) -> Union[ApproximateNegativeSamplingInteractionsDataLoader,
```
Comment (sugi-chan) [inline collie_recs/model/base.py:900]: nice alphabetization... sad bc nicole is leaving

File: collie_recs/model/matrix_factorization.py [hunk 11]
```diff
@@ -57,7 +57,7 @@ class MatrixFactorizationModel(BasePipeline):
         Data loader for training data. If an ``Interactions`` object is supplied, an
         ``InteractionsDataLoader`` will automatically be instantiated with ``shuffle=True``. Note
         that when the model class is saved, datasets will NOT be saved as well
-    vak: ``collie_recs.interactions`` object
+    val: ``collie_recs.interactions`` object
```
Comment (sugi-chan) [inline collie_recs/model/matrix_factorization.py:60]: vak is good too

File: tests/test_model.py [hunk 12]
```diff
@@ -231,6 +236,316 @@ def test_instantiation_of_model_optimizer(train_val_implicit_data):
     trainer_9.fit(model_9)
 
 
+class TestCollieMinimalTrainer():
+    @mock.patch('torch.cuda.is_available')
```
Comment (sugi-chan) [inline tests/test_model.py:240]: yay mocks!

File: tests/test_model.py [hunk 13]
```diff
@@ -231,6 +236,316 @@ def test_instantiation_of_model_optimizer(train_val_implicit_data):
...
+        with pytest.raises(ValueError):
+            trainer.fit(model)
+
+    def test_multiple_optimizers_and_lr_schedulers(self, train_val_implicit_sample_data):
+        train, val = train_val_implicit_sample_data
+        model = MatrixFactorizationModel(train=train,
+                                         val=val,
+                                         optimizer='adam',
+                                         bias_optimizer='sgd',
+                                         lr_scheduler_func=partial(StepLR, step_size=1))
```
Comment (sugi-chan) [inline tests/test_model.py:471]: so in this case the `steplr` gets applied to both adam and the sgd optimizers?

Comment (sugi-chan) [reply tests/test_model.py:471]: what does it look like if you wanted to set 2 seperate lr schedulers? `((scheduler_1,some params),(scheduler_2,some params))`?

**Review by sugi-chan — APPROVED** [review]
I think things look cool and it seems like a good training pipeline with some nifty features! gogo

image

**Comment by nathancooperjones** [comment]
> I think things look cool and it seems like a good training pipeline with some nifty features! gogo
> 
> image

Michael... what is this image????

File: tests/test_movielens.py [hunk 14]
```diff
@@ -129,7 +129,7 @@ def test_run_get_recommendation_visualizations_not_decremented(
             df_user=movielens_explicit_df,
         )
 
-    movielens_df_item_not_decremented = movielens_df_item
+    movielens_df_item_not_decremented = movielens_df_item.copy()
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply tests/test_movielens.py:132]: I guess so? I feel like I should know better by now, but ¯\\_(ツ)_/¯  

Adding the `.copy()` left the original unmodified, so I guess this should have always been there.

File: tests/test_model.py (hunk 13 above)
Comment (nathancooperjones) [reply tests/test_model.py:471]: Hmmm.... that's actually a good idea. Right now, we only accept one and then just use that for each optimizer separately. It might make sense to also have a `bias_lr_scheduler` argument in the future... 🤔 

Good `help wanted` Issue right there!!!!
